### PR TITLE
MQE: Add support for irate and idelta function 

### PR DESCRIPTION
**Pull request by lamida**
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/operators/functions/range_vectors.go [hunk 1]
```diff
@@ -533,3 +533,61 @@ func linearRegression(head, tail []promql.FPoint, interceptTime int64) (slope, i
...
+
+		// We need at least two samples to calculate irate or idelta.
+		if len(fHead)+len(fTail) < 2 {
+			return 0, false, nil, nil
+		}
+
+		var lastSample promql.FPoint
+		var previousSample promql.FPoint
+
+		// If tails has more than two samples, we should use the last two samples from tail.
```
Comment (jhesketh) [inline pkg/streamingpromql/operators/functions/range_vectors.go:559]: (nit)
```suggestion
		// If tail has more than two samples, we should use the last two samples from tail.
```

File: pkg/streamingpromql/operators/functions/range_vectors.go [hunk 2]
```diff
@@ -533,3 +533,61 @@ func linearRegression(head, tail []promql.FPoint, interceptTime int64) (slope, i
...
+	SeriesMetadataFunction: DropSeriesName,
+	StepFunc:               irate(true),
+}
+
+var Idelta = FunctionOverRangeVectorDefinition{
+	SeriesMetadataFunction: DropSeriesName,
+	StepFunc:               irate(false),
+}
+
+func irate(isRate bool) RangeVectorStepFunction {
```
Comment (jhesketh) [inline pkg/streamingpromql/operators/functions/range_vectors.go:547]: (nit)
```suggestion
func irateIdelta(isRate bool) RangeVectorStepFunction {
```

File: pkg/streamingpromql/operators/functions/range_vectors.go [hunk 3]
```diff
@@ -533,3 +533,61 @@ func linearRegression(head, tail []promql.FPoint, interceptTime int64) (slope, i
...
+	return func(step *types.RangeVectorStepData, _ float64, _ types.EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
+		fHead, fTail := step.Floats.UnsafePoints()
+
+		// We need at least two samples to calculate irate or idelta.
+		if len(fHead)+len(fTail) < 2 {
```
Comment (jhesketh) [inline pkg/streamingpromql/operators/functions/range_vectors.go:552]: (nit) `len(fTail)` (and head) are used a lot below. Consider moving to a variable.

File: pkg/streamingpromql/operators/functions/range_vectors.go [hunk 4]
```diff
@@ -533,3 +533,61 @@ func linearRegression(head, tail []promql.FPoint, interceptTime int64) (slope, i
...
+	StepFunc:               irate(true),
+}
+
+var Idelta = FunctionOverRangeVectorDefinition{
+	SeriesMetadataFunction: DropSeriesName,
+	StepFunc:               irate(false),
+}
+
+func irate(isRate bool) RangeVectorStepFunction {
+	return func(step *types.RangeVectorStepData, _ float64, _ types.EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
```
Comment (jhesketh) [inline pkg/streamingpromql/operators/functions/range_vectors.go:548]: (nit)
```suggestion
	return func(step *types.RangeVectorStepData, _ float64, _ types.EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
		// Histograms are ignored
```

**Review by jhesketh — COMMENTED** [review]
Suggestion for some tests:
 - Something with a reset counter
 - Previous sample is a Histogram, so it needs to look back further (if that's consistent with promql)

File: pkg/streamingpromql/operators/functions/range_vectors.go (hunk 3 above)
Comment (lamida) [reply pkg/streamingpromql/operators/functions/range_vectors.go:553]: Addressed in 5383c9cc0dcf10fd046b00314d929a42f753b263

File: pkg/streamingpromql/operators/functions/range_vectors.go (hunk 2 above)
Comment (lamida) [reply pkg/streamingpromql/operators/functions/range_vectors.go]: Addressed in 5383c9cc0dcf10fd046b00314d929a42f753b263

File: pkg/streamingpromql/operators/functions/range_vectors.go (hunk 3 above)
Comment (jhesketh) [reply pkg/streamingpromql/operators/functions/range_vectors.go:553]: Suggest to move the lenTail/lenHead to above this line so you can use it here too

**Comment by lamida** [comment]
> Suggestion for some tests:
> 
> * Something with a reset counter
> * Previous sample is a Histogram, so it needs to look back further (if that's consistent with promql)

More tests are added in edaf8243189fb11d01d550c07bb7ca9c24484659

File: pkg/streamingpromql/operators/functions/range_vectors.go (hunk 3 above)
Comment (lamida) [reply pkg/streamingpromql/operators/functions/range_vectors.go]: Ahh you are right. Moved up in 1b1f3591afcf6d12df817ae972a4fb5e0cc4e6c9

**Review by jhesketh — APPROVED** [review]
LGTM, thanks for that.

We could add a test (outside of the testdata tests) to check the head/tail boundaries of the ring buffer, but I'm happy as it is.
